File: LEETCODE/medium/Subsequence/3404.CountSpecialSubsequence.py
from math import gcd
from collections import defaultdict
from typing import List

class Solution:
    def numberOfSubsequences(self, nums: List[int]) -> int:
        n = len(nums)
        count = defaultdict(int)
        ans = 0

        # Iterate r. For each r, we'll:
        #   1. Add all valid (p, q) pairs where q = r - 2 into the counter
        #   2. Look up matching (s, r) pairs for all valid s
        for r in range(4, n - 2):
            # Add (p, q) pairs where q is fixed at r - 2
            q = r - 2
            for p in range(q - 1):     # p < q - 1, i.e. p <= q - 2
                g = gcd(nums[p], nums[q])
                key = (nums[p] // g, nums[q] // g)
                count[key] += 1

            # Look up (s, r) pairs
            for s in range(r + 2, n):
                g = gcd(nums[s], nums[r])
                key = (nums[s] // g, nums[r] // g)
                ans += count[key]

        return ans


# Tests
print(Solution().numberOfSubsequences([1, 2, 3, 4, 3, 6, 1]))         # 1
print(Solution().numberOfSubsequences([3, 4, 3, 4, 3, 4, 3, 4]))      # 3

File: LEETCODE/medium/Subsequence/3404.CountSpecialSubsequence.js
var numberOfSubsequences = function(nums) {
    const n = nums.length;
    const mp = new Map();
    let ans = 0;

    for (let r = 4; r < n; r++) {
        for (let p = 0, q = r - 2; p < q - 1; p++) {
            const ratio = nums[p] / nums[q];
            mp.set(ratio, (mp.get(ratio) || 0) + 1);
        }
        for (let s = r + 2; s < n; s++) {
            const ratio = nums[s] / nums[r];
            ans += mp.get(ratio) || 0;
        }
    }

    return ans;
};


console.log(numberOfSubsequences([3,4,3,4,3,4,3,4]));